db/corruption_test.cc

```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+  options.wal_recovery_mode = WALRecoveryMode::kPointInTimeRecovery;
+  options.avoid_flush_during_recovery = false;
+  options.env = env_;
+  ASSERT_OK(DestroyDB(dbname_, options));
+  options.create_if_missing = true;
+
+  Reopen(&options);
+  Status s;
+  const std::string test_cf_name = "test_cf";
+  ColumnFamilyHandle* cfh = nullptr;
+  s = db_->CreateColumnFamily(options, test_cf_name, &cfh);
+  ASSERT_OK(s);
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
+  delete cfh;
+  CloseDb();
+
+  std::vector<ColumnFamilyDescriptor> cf_descs;
+  cf_descs.emplace_back(kDefaultColumnFamilyName, options);
+  cf_descs.emplace_back(test_cf_name, options);
+  std::vector<ColumnFamilyHandle*> handles;
+
+  ASSERT_OK(DB::Open(options, dbname_, cf_descs, &handles, &db_));
+
+  // Write one key to test_cf.
+  ASSERT_OK(db_->Put(WriteOptions(), handles[1], "old_key", "dontcare"));
+
+  // Write to default_cf and flush this cf several times to advance wal
+  // number.
+  for (int i = 0; i < 2; ++i) {
+    ASSERT_OK(db_->Put(WriteOptions(), "key" + std::to_string(i), "value"));
+    ASSERT_OK(db_->Flush(FlushOptions()));
+  }
+
+  auto* dbimpl = static_cast_with_check<DBImpl>(db_);
+  assert(dbimpl);
+
+  ASSERT_OK(db_->Put(WriteOptions(), handles[1], "dontcare", "dontcare"));
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
+
+  std::vector<std::string> files;
+  ASSERT_OK(env_->GetChildren(dbname_, &files));
+  uint64_t max_wal_num = 0;
+  for (const auto& f : files) {
+    uint64_t file_num = 0;
+    FileType type{kWalFile};
+    if (ParseFileName(f, &file_num, &type) && type == kWalFile) {
+      if (file_num > max_wal_num) {
+        max_wal_num = file_num;
+      }
+    }
+  }
+
+  for (auto* h : handles) {
+    delete h;
+  }
+  handles.clear();
+  CloseDb();
+
+  std::string wal_path = LogFileName(dbname_, max_wal_num);
+  uint64_t old_size = 0;
+  ASSERT_OK(env_->GetFileSize(wal_path, &old_size));
+  assert(old_size > 8);
+  uint64_t new_size = old_size - 8;
+  ASSERT_OK(test::TruncateFile(env_, wal_path, new_size));
+
+  // Reopen db after first corruption. Default family has higher log number than
+  // corrupted wal number.
+  options.avoid_flush_during_recovery = avoid_flush_during_recovery_;
+  s = DB::Open(options, dbname_, cf_descs, &handles, &db_);
+  ASSERT_OK(s);
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
+  for (auto* h : handles) {
+    delete h;
+  }
+  handles.clear();
+  CloseDb();
+
+  max_wal_num = 0;
+  files.clear();
+  ASSERT_OK(env_->GetChildren(dbname_, &files));
+  for (const auto& f : files) {
+    uint64_t file_num = 0;
+    FileType type{kWalFile};
+    if (ParseFileName(f, &file_num, &type) && type == kWalFile) {
+      if (file_num > max_wal_num) {
+        max_wal_num = file_num;
+      }
+    }
+  }
+  wal_path = LogFileName(dbname_, max_wal_num);
+  ASSERT_OK(test::TruncateFile(env_, wal_path, new_size = 0));
```

Comment:
Regarding this block together with the above `DB::Open()`, I think it's helpful to add comments and/or put it in a helper function. The comment should explain what failure case we are emulating: before this PR, we may prematurely persist a new MANIFEST even before we can declare the DB is in consistent state after recovery (this is when the new WAL is synced). Should there be a power failure before we sync the new WAL, we will end up in a situation in which the MANIFEST indicates one or more column family's log number is ahead of the WAL with corruption. Also helpful to explain briefly why the corrupted WAL will still be examined in the next recovery attempt.

Reply:
Same for the other unit test of transaction db.

Reply:
Sure. I will update the comments.